bin/propolis-server/src/lib/serial/backend.rs

```diff
@@ -0,0 +1,397 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A backend that provides external clients with access to a Propolis character
+//! device and tracks the history of bytes the guest has written to that device.
+
+use std::{
+    collections::BTreeMap,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+};
+
+use propolis::chardev::{
+    pollers::{SinkBuffer, SourceBuffer},
+    Sink, Source,
+};
+use tokio::{
+    select,
+    sync::{mpsc, oneshot},
+};
+
+use super::history_buffer::{HistoryBuffer, SerialHistoryOffset};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct ClientId(u64);
+
+/// A client's rights when accessing a character backend.
+#[derive(Clone, Copy)]
+pub(super) enum Permissions {
+    /// The client may both read and write to the backend.
+    ReadWrite,
+
+    /// The client may only read from the backend.
+    ReadOnly,
+}
+
+impl Permissions {
+    pub(super) fn is_writable(&self) -> bool {
+        matches!(self, Permissions::ReadWrite)
+    }
+}
+
+/// Determines what happens when the backend is unable to send a guest byte back
+/// to a client because the client's channel is full.
+#[derive(Clone, Copy)]
+pub(super) enum FullReadChannelDiscipline {
+    /// The backend should block until it can send to this client.
+    Block,
+
+    /// The backend should close this client's connection.
+    Close,
+}
+
+/// An individual client of a character backend.
+struct Client {
+    /// Bytes read from the character device should be sent to the client on
+    /// this channel.
+    tx: mpsc::Sender<u8>,
```

Comment:
`mpsc::Sender<u8>` is a pretty unfortunate way to represent an asynchronous stream of bytes, from a performance standpoint: this will allocate a node in a linked list for each individual byte. If you want an asynchronous byte pipe, I would consider using something like [`tokio::io::simplex`](https://docs.rs/tokio/latest/tokio/io/fn.simplex.html) , which is an asynchronous byte pipe implemented as a single contiguous in memory buffer. There's also [`DuplexStream`](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html), which is the same thing but bi-directional.

Reply:
`simplex` definitely seems like a much nicer tool here. The one gotcha with it is that `Sender` is `Clone`, but `SimplexStream` and its `WriteHalf` aren't, so the code in `read_task` that captures channels under the lock and then drops the lock before sending (so that it can guiltlessly block as needed) won't work as-is.

Now that I think about it, I wonder if it'd be better to rewrite `read_task` so that it doesn't need to look at the client map at all. (I think there's even an outside chance I could get rid of the client map altogether...) I'm going to give this a try to see how it shakes out.